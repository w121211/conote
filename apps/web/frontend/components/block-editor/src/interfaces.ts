import type { Token } from 'prismjs'
import type {
  NoteDocContentBodyInput,
  NoteDocContentHeadInput,
} from 'graphql-let/__generated__/__types__'
import type {
  NoteDraftEntryFragment,
  NoteDraftFragment,
  NoteFragment,
  SearchHitFragment,
} from '../../../../apollo/query.graphql'

//
// Component State - commonly used in various components & handlers
//
//
//
//
//
//

export type CaretPosition = {
  top: number
  left: number
  height: number
}

export type DragTarget = 'first' | 'before' | 'after'

/**
 * athens naming:
 * - term -> query
 * - hit-index -> index
 * - hits -> results
 */
export type Search = {
  type: SearchType | null
  term: string | null
  hitIndex: number | null
  // hits: SearchHit[]
  hits: SearchHitFragment[]
}

// export type SearchHit = Omit<SearchHitFragment, '__typename'> & {
//   discussTitle?: string
//   note?: {
//     symbol: string
//     webpageTitle?: string
//   }
// }

export type SearchType = 'topic' | 'discuss' // | 'slash' | 'hashtag' | 'template'

//
// Events & Dom events
//
//
//
//
//
//

export type DestructTextareaKeyEvent = {
  value: string
  start: number
  end: number
  head: string
  tail: string
  selection: string
  key: string // keyboard key
  keyCode: number
  target: HTMLTextAreaElement
  shift: boolean
  meta: boolean
  ctrl: boolean
  alt: boolean
}

/**
 * A block's relative position, defined by a reference block / a doc
 *
 * @param refBlockUid ref-block uid
 * @param docTitle if ref-block is a doc, use doc-title instead of block-uid
 * @param relation position relate to ref-block
 */
export type BlockPosition = {
  refBlockUid?: string

  // TODO: Consider to remove?
  docTitle?: string

  relation: BlockPositionRelation
}

/**
 * parent-child relation: 'first' | 'last'
 * siblings relation: 'before' | 'after'
 */
export type BlockPositionRelation = 'first' | 'last' | 'before' | 'after'

//
// Stores
//
//
//
//
//
//

// export type NodeProps = {
//   db: {
//     id?: string
//   }
//   block: {
//     uid?: string
//     children?: { uid: string }[]
//   }
//   node: {
//     title?: string
//   }
//   page: {
//     sidebar?: false
//   }
// }

// export type NodeElProps = {
//   menu: {
//     show: boolean
//   }
//   title: {
//     initial?: string
//     local?: string
//   }
//   alert: {
//     show?: null
//     message?: null
//     confirmFn?: null
//     cancelFn?: null
//   }
//   // LinkedReferences: true
//   // UnlinkedReferences: false
//   node?: NodeProps
// }

export type Block = {
  uid: string

  // Null for doc-block
  parentUid: string | null
  childrenUids: string[]
  order: number

  // Doc block only, equals to doc.str
  docSymbol?: string

  str: string
  open?: boolean

  // TBC, consider to drop
  editTime?: number
}

export type BlockWithChildren = Block & { children: BlockWithChildren[] }

/**
 * In athensresearch, 'Doc' is named as 'Node' and 'Page',
 * alone with a concept of node-block (node-page-block), ndoe-title, page-title, page-block (or context-block)
 */
export type Doc = {
  // Client side only id, generate one when first open the doc
  uid: string

  // Should be unique most of time except when renaming
  symbol: string

  branch: string

  domain: string

  // Note-doc-content-head is editable and stores here
  contentHead: NoteDocContentHeadInput

  // Use 'input' instead of 'fragment' for updating the note-draft
  contentBody: Omit<NoteDocContentBodyInput, 'blocks'>

  // Refer to root-block-uid
  blockUid: string

  // The note & head-doc that used to create the draft, ie the from-doc
  noteCopy: Omit<NoteFragment, '__typename'> | null

  // The current saved note-draft
  noteDraftCopy: Omit<NoteDraftFragment, '__typename'>
}

/**
 *
 */
export type CursorProp = {
  cursor: {
    blockUid: string
    caretIndex: number
  } | null
}

/**
 * The final single stop to store all required info for the editor
 *
 * One browser window can only have one editor, an editor includes
 * - route
 * - one main-doc
 * - one modal-doc (optional)
 * - one sidebar
 *   - all editing draft-entries
 * - alert
 */
export type EditorProps = {
  alert: {
    show?: null
    message?: null
    confirmFn?: null
    cancelFn?: null
  }
  leftSidebar: {
    show: boolean
    droppedItems: NoteDraftEntryFragment[]
    editingItems: NoteDraftEntryFragment[]
  }
  modal: {
    discuss?: {
      id?: string
      title: string
    }
    doc?: {
      title: string
    }
    mainEditorCursor?: {
      blockUid: string
      anchor: number
      offset: number
    }
  }
  opening: {
    main: {
      symbol: string | null
      docUid: string | null
    }
    modal: {
      symbol: string | null
      docUid: string | null
    }

    // (Future) when open a block
    // blockUid?: string
  }

  // Experimental
  draftEntries: NoteDraftEntryFragment[]
  chains: NoteDraftEntryFragment[][]

  // Only allow to open one tab, so no need to be a list
  tab: {
    chain: {
      entry: NoteDraftEntryFragment
      docUid: string
    }[]
    // Modify by chainItemOpen event
    openDraftId: string | null
    //
    loading: boolean
  }
}

//
// Inline Items
//
//
//
//
//
//

export interface InlineBase {
  type: string
  str: string
  token: Token
}

export interface InlineText extends InlineBase {
  type: 'text'
}

// export type InlineAuthor = {
//   type: 'author'
//   str: string
//   authorName: string
// }

export interface InlineDiscuss extends InlineBase {
  type: 'inline-discuss'
  title: string
  id?: string
}

export interface InlineFiltertag extends InlineBase {
  type: 'inline-filtertag'
}

// export type InlineMirror = {
//   type: 'inline-mirror'
//   str: string
//   // symbolType: SymbolType
//   mirrorSymbol: string
//   author?: string // 需要 parse 取得，eg ::$XX @cnyes
// }

// export type InlineEmoji = {
//   type: 'emoji'
//   str: string
//   id: number
//   hashtag?: GQLHashtag
// }

// export type InlineNewHashtag = {
//   type: 'new-hashtag'
//   str: string
// }

export interface InlinePoll extends InlineBase {
  type: 'inline-poll'
  id?: string
  choices: string[]
  // pollCopy?: PollFragment
  // vote?: author vote
}

export interface InlineRate extends InlineBase {
  type: 'inline-rate'
  id?: string
  params: string[]
  // rateCopy?: RateFragment
  authorName?: string
  targetSymbol?: string
  // choice?: RateChoice
  // vote?: author vote
}

export interface InlineSymbol extends InlineBase {
  type: 'inline-symbol'
  // symbolType: SymbolType
  symbol: string
}

export interface InlineComment extends InlineBase {
  type: 'inline-comment'
}

export type InlineItem =
  | InlineText
  // | InlineAuthor
  | InlineDiscuss
  | InlineFiltertag
  // | InlineMirror
  // | InlineHashtag
  // | InlineNewHashtag
  | InlinePoll
  | InlineRate
  | InlineSymbol
  | InlineComment

function isInlinePoll(item: InlineItem): item is InlinePoll {
  return item.type === 'inline-poll'
}

function isInlineRate(item: InlineItem): item is InlineRate {
  return item.type === 'inline-rate'
}
