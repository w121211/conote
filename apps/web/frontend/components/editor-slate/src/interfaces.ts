import type { TreeNodeChangeType } from '@conote/docdiff'
import type { BaseRange } from 'slate'
import type {
  Block,
  InlineComment,
  InlineDiscuss,
  InlineFiltertag,
  InlineItem,
  InlinePoll,
  InlineRate,
  InlineSymbol,
} from '../../editor-textarea/src/interfaces'
import { IndenterFormatErrorCode } from './indenter/normalizers'

//
// Component props
//
//
//
//
//
//

export type LeafPopoverProps<T extends InlineItem> = {
  // Element's dom id, used to identify the caret sitting (on-select-event) element
  id: string

  blockUid: string
  docUid: string
  draftId: string
  inlineItem: T
}

//
// Slate types
//
//
//
//
//
//

export type ElementIndenter = {
  type: 'indenter'
  children: TextCustom[]

  // Indicate the current level of indent, start from 1, since 0 represent no indent
  indent: number

  errorCode?: IndenterFormatErrorCode

  // Client-side id
  uid: string

  // Keep the original copy
  blockCopy?: Omit<Block, 'childrenUids'>

  // Record change-event lively
  change?: TreeNodeChangeType

  // For markdown render
  markdownIndent?: number
}

export type ElementLic = {
  type: 'lic'
  children: (TextCustom | InlineElementCustom)[]

  // Client-side id
  uid: string

  // Keep the original copy
  blockCopy?: Block

  // Record change-event lively
  change?: TreeNodeChangeType

  // editingBody?: true
  // selected?: boolean
  // asAuthor?: true
  // banAsOauthor?: true
  // banDeleteBackward?: true
  // banDeleteForward?: true
  // banInsertBreak?: true
  // insertBreakAsIndent?: true
}

export type ElementLi = {
  type: 'li'
  children: [ElementLic, ElementUl?]
}

export type ElementUl = {
  type: 'ul'
  children: ElementLi[]
  folded?: true
}

export type ElementCustom =
  // | ElementLic
  // | ElementLi
  // | ElementUl
  // | InlineElementCustom
  ElementIndenter

export type InlineElementDiscuss = InlineDiscuss & {
  children: TextCustom[]
  // children: Descendant[]
}

export type InlineElementFiltertag = InlineFiltertag & {
  children: TextCustom[]
}

// export type InlineMirrorElement = InlineMirror & {
//   children: CustomText[]
// }

export type InlineElementPoll = InlinePoll & {
  children: TextCustom[]
  selected?: boolean
}

export type InlineElementRate = InlineRate & {
  children: TextCustom[]
}

export type InlineElementSymbol = InlineSymbol & {
  children: TextCustom[]
}

export type InlineElementComment = InlineComment & {
  children: TextCustom[]
}

export type InlineElementCustom =
  | InlineElementDiscuss
  | InlineElementFiltertag
  // | InlineMirrorElement
  | InlineElementPoll
  | InlineElementRate
  | InlineElementSymbol
  | InlineElementComment

export type RangeCustom = BaseRange & {
  blockUid?: string
  inlineItem?: InlineItem
}

export type TextCustom = {
  text: string

  bold?: true
  italic?: true
  code?: true

  // Received from 'RangeCustom' during decorate
  blockUid?: string
  inlineItem?: InlineItem
}

export type TextEmpty = {
  text: string
}
