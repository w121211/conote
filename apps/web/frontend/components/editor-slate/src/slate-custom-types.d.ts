import type { BaseEditor } from 'slate'
import type { ReactEditor } from 'slate-react'
import type { HistoryEditor } from 'slate-history'
import type { ElementCustom, RangeCustom, TextCustom } from './interfaces'

type CustomEditor = BaseEditor & ReactEditor & HistoryEditor

declare module 'slate' {
  interface CustomTypes {
    Editor: CustomEditor
    Element: ElementCustom
    Text: TextCustom
    Range: RangeCustom
  }
}
