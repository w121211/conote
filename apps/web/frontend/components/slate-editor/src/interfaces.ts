import type { TreeNodeChangeType } from '@conote/docdiff'
import type {
  Block,
  InlineComment,
  InlineDiscuss,
  InlineFiltertag,
  InlinePoll,
  InlineRate,
  InlineSymbol,
} from '../../block-editor/src/interfaces'

//
// Slate element
//
//
//
//
//
//

export type ElementLc = {
  type: 'lc'
  children: (TextCustom | InlineElementCustom)[]

  // Client-side id
  uid: string

  // Keep the original copy
  blockCopy?: Block

  // Record change-event lively
  change?: TreeNodeChangeType

  // editingBody?: true
  // selected?: boolean
  // asAuthor?: true // 若沒有的話視為self author
  // banAsOauthor?: true // 此欄位無法以 @oauthor 記錄，例如self card
  // banDeleteBackward?: true
  // banDeleteForward?: true
  // banInsertBreak?: true
  // insertBreakAsIndent?: true
}

export type ElementLi = {
  type: 'li'
  children: [ElementLc, ElementUl?]
}

export type ElementUl = {
  type: 'ul'
  children: ElementLi[]
  folded?: true
}

//
//
//
//
//
//
//
//

export type InlineElementDiscuss = InlineDiscuss & {
  children: TextCustom[]
  // children: Descendant[]
}

export type InlineElementFiltertag = InlineFiltertag & {
  children: TextCustom[]
}

// export type InlineMirrorElement = InlineMirror & {
//   children: CustomText[]
// }

export type InlineElementPoll = InlinePoll & {
  children: TextCustom[]
  selected?: boolean
}

export type InlineElementRate = InlineRate & {
  children: TextCustom[]
}

export type InlineElementSymbol = InlineSymbol & {
  children: TextCustom[]
}

export type InlineElementComment = InlineComment & {
  children: TextCustom[]
}

export type InlineElementCustom =
  | InlineElementDiscuss
  | InlineElementFiltertag
  // | InlineMirrorElement
  | InlineElementPoll
  | InlineElementRate
  | InlineElementSymbol
  | InlineElementComment

//
//
//
//
//
//
//
//

export type TextCustom = {
  bold?: true
  italic?: true
  code?: true
  text: string

  // Stream type
  type?: string

  tokenType?: string
  // placeholder?: boolean
}

export type TextEmpty = {
  text: string
}
