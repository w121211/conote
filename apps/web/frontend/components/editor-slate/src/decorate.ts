/**
 * Mainly copy from: https://github.com/ianstormtaylor/slate/blob/main/site/examples/code-highlighting.tsx
 */
import { Editor, Element, NodeEntry, Text } from 'slate'
import { parseBlockString } from '../../editor-textarea/src/parse-render'
import type { ElementIndenter, RangeCustom } from './interfaces'

export const decorate = (
  [node, path]: NodeEntry,
  editor: Editor,
): RangeCustom[] => {
  const ranges: RangeCustom[] = []

  if (!Text.isText(node)) return ranges

  const entry = Editor.above<ElementIndenter>(editor, {
    match: n => Element.isElementType<ElementIndenter>(n, 'indenter'),
    at: path,
  })
  if (entry === undefined) throw new Error('Indenter not found.')

  const { inlineItems } = parseBlockString(node.text),
    [n] = entry,
    { uid: blockUid } = n

  let start = 0
  for (const e of inlineItems) {
    const end = start + e.str.length
    ranges.push({
      anchor: { path, offset: start },
      focus: { path, offset: end },
      inlineItem: e,
      blockUid,
    })
    // console.log(e.type, e.token)
    start = end
  }

  return ranges
}
