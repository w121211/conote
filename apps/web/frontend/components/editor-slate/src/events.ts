import type { NoteDraftInput } from 'graphql-let/__generated__/__types__'
import { differenceBlocks, parseGQLBlocks } from '../../../../share/utils'
import type { Block, Doc } from '../../editor-textarea/src/interfaces'
import { docUpdatePropsOp } from '../../editor-textarea/src/op/ops'
import { noteDraftService } from '../../editor-textarea/src/services/note-draft.service'
import { docRepo } from '../../editor-textarea/src/stores/doc.repository'
import {
  BlockInput,
  writeBlocks,
} from '../../editor-textarea/src/utils/block-writer'
import { blocksToIndenters, indentersToBlocks } from './indenter/serializers'
import type { ElementIndenter } from './interfaces'
// import { blocksToLiList, elementLisToBlocks } from './serializers'
import { docValueRepo } from './stores/doc-value.repository'
import { genTemplateBlocks } from './templates'

function toGQLBlocks(blocks: Block[]) {
  return blocks.map(e => {
    const { editTime, childrenUids, ...rest } = e
    return { ...rest }
  })
}

export function toNoteDraftInput(
  doc: Doc,
  value: ElementIndenter[],
): NoteDraftInput {
  const { blockUid: docBlockUid, noteCopy, noteDraftCopy, contentHead } = doc,
    docBlock = docRepo.getDocBlock(doc),
    bodyBlocks = indentersToBlocks(value, docBlockUid),
    bodyBlocks_: Block[] = bodyBlocks.map(e => ({ ...e, childrenUids: [] })),
    blocks = [docBlock, ...bodyBlocks_],
    blocks_ = toGQLBlocks(blocks)

  console.debug(blocks_)

  // Null is a valide input for differencer, but '[]' is not
  const startBlocks = noteCopy
      ? noteCopy.headDoc.contentBody.blocks.map(e => ({
          ...e,
          parentUid: e.parentUid ?? null,
          open: e.open ?? undefined,
          docSymbol: e.docSymbol ?? undefined,
        }))
      : null,
    blockDiff = differenceBlocks(blocks_, startBlocks),
    input: NoteDraftInput = {
      fromDocId: noteCopy?.headDoc.id,
      domain: noteDraftCopy.domain,
      contentHead,
      contentBody: {
        // discussIds: [],
        // symbols: [],
        blocks: blocks_,
        blockDiff: blockDiff,
      },
    }
  return input
}

function isValueChanged(docUid: string, value: ElementIndenter[]) {
  const doc = docRepo.getDoc(docUid),
    { noteDraftCopy } = doc

  const input = toNoteDraftInput(doc, value),
    { blocks: startBlocks } = parseGQLBlocks(noteDraftCopy.contentBody.blocks),
    { blocks: finalBlocks } = parseGQLBlocks(input.contentBody.blocks)

  // console.debug(finalBlocks, startBlocks)

  const changes = differenceBlocks(finalBlocks, startBlocks),
    changed = changes.length > 0

  return { changed, changes, doc, noteDraftCopy, input }
}

/**
 * @returns true if doc is saved, returns false if doc is not perform save (eg, content not change)
 *
 * @throws Value validation error, eg 'indent_oversize'
 */
export async function slateDocSave(
  docUid: string,
  newSymbol?: string,
  opts = {
    force: false,
  },
) {
  const { value } = docValueRepo.getDocValue(docUid)
  const { changed, input, noteDraftCopy } = isValueChanged(docUid, value)

  if (opts.force || changed) {
    const draft = await noteDraftService.updateDraft(
        noteDraftCopy.id,
        input,
        newSymbol,
      ),
      op = docUpdatePropsOp(docUid, { noteDraftCopy: draft })

    docRepo.update(op)
    return true
  }

  return false
}

export function docTemplateGenerate(doc: Doc, title: string) {
  const docBlock = docRepo.getDocBlock(doc),
    blocks = genTemplateBlocks(title.toLowerCase(), docBlock),
    indenters = blocksToIndenters(blocks)

  return indenters
}
