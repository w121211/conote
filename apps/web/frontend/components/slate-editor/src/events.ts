import type { NoteDraftInput } from 'graphql-let/__generated__/__types__'
import { differenceBlocks, parseGQLBlocks } from '../../../../share/utils'
import type { Block, Doc } from '../../block-editor/src/interfaces'
import { docUpdatePropsOp } from '../../block-editor/src/op/ops'
import { noteDraftService } from '../../block-editor/src/services/note-draft.service'
import { docRepo } from '../../block-editor/src/stores/doc.repository'
import {
  BlockInput,
  writeBlocks,
} from '../../block-editor/src/utils/block-writer'
import type { ElementLi } from './interfaces'
import { blocksToLiList, elementLisToBlocks } from './serializers'
import { docValueRepo } from './stores/doc-value.repository'

function toGQLBlocks(blocks: Block[]) {
  return blocks.map(e => {
    const { editTime, childrenUids, ...rest } = e
    return { ...rest }
  })
}

function toNoteDraftInput(doc: Doc, value: ElementLi[]): NoteDraftInput {
  const { blockUid, noteCopy, noteDraftCopy, contentHead } = doc,
    docBlock = docRepo.getDocBlock(doc),
    bodyBlocks = elementLisToBlocks(value, blockUid),
    blocks = [docBlock, ...bodyBlocks],
    blocks_ = toGQLBlocks(blocks)

  // Null is a valide input for differencer, but '[]' is not
  const startBlocks = noteCopy
      ? noteCopy.headDoc.contentBody.blocks.map(e => ({
          ...e,
          parentUid: e.parentUid ?? null,
          open: e.open ?? undefined,
          docSymbol: e.docSymbol ?? undefined,
        }))
      : null,
    blockDiff = differenceBlocks(blocks_, startBlocks),
    input: NoteDraftInput = {
      fromDocId: noteCopy?.headDoc.id,
      domain: noteDraftCopy.domain,
      contentHead,
      contentBody: {
        // discussIds: [],
        // symbols: [],
        blocks: blocks_,
        blockDiff: blockDiff,
      },
    }
  return input
}

function isValueChanged(docUid: string, value: ElementLi[]) {
  const doc = docRepo.getDoc(docUid),
    { noteDraftCopy } = doc

  const input = toNoteDraftInput(doc, value),
    { blocks: startBlocks } = parseGQLBlocks(noteDraftCopy.contentBody.blocks),
    { blocks: finalBlocks } = parseGQLBlocks(input.contentBody.blocks),
    changes = differenceBlocks(finalBlocks, startBlocks),
    changed = changes.length > 0

  return { changed, changes, doc, noteDraftCopy, input }
}

export async function slateDocSave(
  docUid: string,
  // value: ElementLi[],
  newSymbol?: string,
  opts = {
    force: false,
  },
) {
  const { value } = docValueRepo.getDocValue(docUid)

  const { changed, input, noteDraftCopy } = isValueChanged(docUid, value)
  // console.debug('docSave', changed, value)

  if (opts.force || changed) {
    const draft = await noteDraftService.updateDraft(
        noteDraftCopy.id,
        input,
        newSymbol,
      ),
      op = docUpdatePropsOp(docUid, { noteDraftCopy: draft })

    docRepo.update(op)
    // console.debug('docRepo.update')
  }
}

function genTemplateGeneral(): BlockInput {
  const general: BlockInput = [
    'DUMMY_ROOT',
    [
      [
        'Basic',
        [
          ['What is ...?', ['']],
          // ['Explain ... in one picture?', ['']],
          ['Why and why not ...?', ['']],
          ['How ...?', ['']],
          // ['How this research solve?', ['']],
          ['What difference compared to the others?', ['']],
        ],
      ],
      ['Discuss', ['']],
    ],
  ]
  return general
}

export function docTemplateGenerate(doc: Doc) {
  const docBlock = docRepo.getDocBlock(doc),
    blocks = writeBlocks(genTemplateGeneral(), { docBlock })
  return blocksToLiList(blocks)
}
