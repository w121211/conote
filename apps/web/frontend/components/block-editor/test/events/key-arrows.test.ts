import { addEntities } from '@ngneat/elf-entities'
import { enterNewBlock, keyArrowUp } from '../../src/events'
import { Block } from '../../src/interfaces'
import {
  blockRepo,
  blocksStore,
  getBlock,
} from '../../src/stores/block.repository'
import { rfdbRepo } from '../../src/stores/rfdb.repository'

const blocks: Block[] = [
  {
    uid: '0',
    str: '0',
    order: 0,
    parentUid: null,
    childrenUids: ['1', '2', '3'],
  },
  { uid: '1', str: '1', order: 0, parentUid: '0', childrenUids: [] },
  { uid: '2', str: '2', order: 1, parentUid: '0', childrenUids: [] },
  { uid: '3', str: '3', order: 2, parentUid: '0', childrenUids: [] },
]

beforeEach(() => {
  blockRepo.update([addEntities(blocks)])
})

it('enterNewBlock()', () => {
  rfdbRepo.setProps({ editing: { uid: '0' } })

  const parent = blocks[0],
    block = blocks[1]

  // events.up('1', 'end')
  enterNewBlock(block, 'x')
  expect(blocksStore.getValue().entities).toMatchSnapshot()

  // expect(rfdbRepo.getValue().editing?.uid).toEqual('1')
})

it('enterNewBlock, enterNewBlock, up', () => {
  rfdbRepo.setProps({ editing: { uid: '0' } })

  const parent = blocks[0]

  // events.up('1', 'end')
  enterNewBlock(blocks[1], 'x')
  expect(rfdbRepo.getValue().editing?.uid).toEqual('x')

  keyArrowUp('x', 0)
  expect(rfdbRepo.getValue().editing?.uid).toEqual('1')

  enterNewBlock(getBlock('x'), 'y')
  expect(blocksStore.getValue().entities).toMatchSnapshot()
  expect(rfdbRepo.getValue().editing?.uid).toEqual('y')

  keyArrowUp('y', 0)
  expect(rfdbRepo.getValue().editing?.uid).toEqual('x')

  // expect(rfdbRepo.getValue().editing?.uid).toEqual('1')
})

it('up()', () => {
  rfdbRepo.setProps({ editing: { uid: '1' } })
  keyArrowUp('0', 'end')
  expect(rfdbRepo.getValue().editing?.uid).toEqual('0')

  // const x = { uid: 'x', str: 'x', order: 0, parentUid: 'x', childrenUids: [] }

  // expect(insert(v, x, 'first', v[0])).toEqual([x, ...v])
  // expect(insert(v, x, 'first', v[1])).toEqual([x, ...v])
  // expect(insert(v, x, 'first', v[2])).toEqual([x, ...v])

  // expect(insert(v, x, 'last', v[0])).toEqual([...v, x])
  // expect(insert(v, x, 'last', v[1])).toEqual([...v, x])
  // expect(insert(v, x, 'last', v[2])).toEqual([...v, x])

  // // console.log(insert(v, x, 'before', v[0]))
  // expect(insert(v, x, 'before', v[0])).toEqual([x, ...v])
  // expect(insert(v, x, 'before', v[1])).toEqual([v[0], x, v[1], v[2]])
  // expect(insert(v, x, 'before', v[2])).toEqual([v[0], v[1], x, v[2]])

  // expect(insert(v, x, 'after', v[0])).toEqual([v[0], x, v[1], v[2]])
  // expect(insert(v, x, 'after', v[1])).toEqual([v[0], v[1], x, v[2]])
  // expect(insert(v, x, 'after', v[2])).toEqual([...v, x])
})

it('down()', () => {
  rfdbRepo.setProps({ editing: { uid: '1' } })
  keyArrowUp('0', 'end')
  expect(rfdbRepo.getValue().editing?.uid).toEqual('0')

  // const x = { uid: 'x', str: 'x', order: 0, parentUid: 'x', childrenUids: [] }

  // expect(insert(v, x, 'first', v[0])).toEqual([x, ...v])
  // expect(insert(v, x, 'first', v[1])).toEqual([x, ...v])
  // expect(insert(v, x, 'first', v[2])).toEqual([x, ...v])

  // expect(insert(v, x, 'last', v[0])).toEqual([...v, x])
  // expect(insert(v, x, 'last', v[1])).toEqual([...v, x])
  // expect(insert(v, x, 'last', v[2])).toEqual([...v, x])

  // // console.log(insert(v, x, 'before', v[0]))
  // expect(insert(v, x, 'before', v[0])).toEqual([x, ...v])
  // expect(insert(v, x, 'before', v[1])).toEqual([v[0], x, v[1], v[2]])
  // expect(insert(v, x, 'before', v[2])).toEqual([v[0], v[1], x, v[2]])

  // expect(insert(v, x, 'after', v[0])).toEqual([v[0], x, v[1], v[2]])
  // expect(insert(v, x, 'after', v[1])).toEqual([v[0], v[1], x, v[2]])
  // expect(insert(v, x, 'after', v[2])).toEqual([...v, x])
})
