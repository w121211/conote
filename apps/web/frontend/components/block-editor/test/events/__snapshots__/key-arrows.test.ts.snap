// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enterNewBlock() 1`] = `
Object {
  "0": Object {
    "childrenUids": Array [
      "1",
      "x",
      "2",
      "3",
    ],
    "editTime": 1647622347613,
    "order": 0,
    "parentUid": null,
    "str": "0",
    "uid": "0",
  },
  "1": Object {
    "childrenUids": Array [],
    "order": 0,
    "parentUid": "0",
    "str": "1",
    "uid": "1",
  },
  "2": Object {
    "childrenUids": Array [],
    "order": 2,
    "parentUid": "0",
    "str": "2",
    "uid": "2",
  },
  "3": Object {
    "childrenUids": Array [],
    "order": 3,
    "parentUid": "0",
    "str": "3",
    "uid": "3",
  },
  "x": Object {
    "childrenUids": Array [],
    "editTime": 1647622347613,
    "open": true,
    "order": 1,
    "parentUid": "0",
    "str": "",
    "uid": "x",
  },
}
`;

exports[`enterNewBlock, enterNewBlock, up 1`] = `
Object {
  "0": Object {
    "childrenUids": Array [
      "1",
      "x",
      "y",
      "2",
      "3",
    ],
    "editTime": 1647622347619,
    "order": 0,
    "parentUid": null,
    "str": "0",
    "uid": "0",
  },
  "1": Object {
    "childrenUids": Array [],
    "order": 0,
    "parentUid": "0",
    "str": "1",
    "uid": "1",
  },
  "2": Object {
    "childrenUids": Array [],
    "order": 3,
    "parentUid": "0",
    "str": "2",
    "uid": "2",
  },
  "3": Object {
    "childrenUids": Array [],
    "order": 4,
    "parentUid": "0",
    "str": "3",
    "uid": "3",
  },
  "x": Object {
    "childrenUids": Array [],
    "editTime": 1647622347617,
    "open": true,
    "order": 1,
    "parentUid": "0",
    "str": "",
    "uid": "x",
  },
  "y": Object {
    "childrenUids": Array [],
    "editTime": 1647622347619,
    "open": true,
    "order": 2,
    "parentUid": "0",
    "str": "",
    "uid": "y",
  },
}
`;
